### PR TITLE
Added support for generated columns and identity columns

### PR DESCRIPTION
**Pull request by Oreilles**
* Added support for generated columns for all vendors
* Added support for identity columns for Oracle and Postgres

### CODE REVIEW

**Comment by aidenfoxx** [comment]
For Oracle, I'm not sure if we should use `IDENTITY_COLUMN` as a reference for `has_auto_increment`, since a lot of Oracle databases still use triggers for auto incrementing (including any schemas generated by knex). We should at least add a note that the `has_auto_increment` value is unreliable for Oracle.

Also, I tested the new query in Oracle 18c and it's failing. You can simply change the query to use `USER_TAB_COLS` instead of `USER_TAB_COLUMNS` and that should fix the issue.

I recently got the Oracle tests working, so you can update those to verify your changes! 😄

**Comment by Oreilles** [comment]
> Also, I tested the new query in Oracle 18c and it's failing. You can simply change the query to use USER_TAB_COLS instead of USER_TAB_COLUMNS and that should fix the issue.

Should it be replaced everywhere ? And wouldn't it break previous versions ? ( I have zero experience with Oracle databases )

**Comment by aidenfoxx** [comment]
> Should it be replaced everywhere ? And wouldn't it break previous versions ?

You can replace it everywhere. It's just a strange quirk of Oracle.

> an auto increment is actually a type in every other vendor: auto_increment in MySQL, identity or serial in MSSQL and Postgres. Now Oracle has one too so it makes sense to use it.

I get that, but my point is that it's more common in Oracle land (in my experience) that an auto_increment will be handled by a trigger rather than `IDENTITY_COLUMN`. This includes any schema generated by knex.

So with the new change we'll start getting `false` as the `has_auto_increment` value for columns that auto increment via triggers. This is semantically confusing, as the column is auto_incrementing in practice. Currently the `has_auto_increment` value will always be `undefined` which makes it clear to the programmer that it is an unreliable value.

I'm fine with keeping the change, however, I just think we should document that it's unreliable in Oracle.

**Comment by Oreilles** [comment]
That's an easy fix though: `has_auto_increment: 'IDENTITY_COLUMN' == 'yes' || undefined`

**Comment by aidenfoxx** [comment]
That could be the solution. I'm mostly just thinking out loud since I really don't know what is best. 😅

**Comment by Oreilles** [comment]
IMO it's not unreliable: having a trigger function on a field doesn't make it an `auto_increment`. Also it's not like it's going to break anyone's code: people using Oracle database couldn't use that column anyway, and I doubt that they'd have written code that heavily relied on `has_auto_increment === undefined` 😇

Identity columns were introduced 8 years ago in Oracle, and 5 years ago in Postgres. It's probably time for `knex` to use these are they're the recommanded - and SQL compliant - way. Opportunity for another PR ?

**Comment by aidenfoxx** [comment]
As I said, my experience is that most Oracle software I've worked with still uses triggers for auto increments. But I don't consider myself an Oracle expert.

But when I say unreliable, I mean more in the sense that it could be confusing for someone using knex migrations to see that the schema-inspector reports `has_auto_increment = false` after they specifically generated a column to be `auto_increment`. But as you say, that's really not the problem of `knex-schema-inspector`.

I'm happy with the changes, and think we probably should just leave any funky trigger stuff on the consumer. But it might be nice to raise an issue in knex to start using IDENTITY.

**Comment by Oreilles** [comment]
I tried to run the test, works fine in Sqlite but fails in Postgres and MySQL because the table columns are not returned in the order expected.

**Comment by aidenfoxx** [comment]
Oh man, that's annoying. The Postgres and MySQL tests were consistent last time I worked on it, but I had to add sorting to the Oracle tests for that issue. The order would be random. You can probably use the Oracle tests as a reference if you need to fix the tests:

https://github.com/knex/knex-schema-inspector/blob/01929fb70bc2391309a2e283ba5ed198f1c4769a/test/oracle.spec.ts#L97

**Comment by Oreilles** [comment]
Not gonna do that for every test for every vendor 🍉

**Comment by aidenfoxx** [comment]
Fair. As long as the pipeline tests pass on github it doesn't matter too much. 👍 

Just need someone to approve that. Paging @rijkvanzanten.

File: lib/types/column.ts [hunk 1]
```diff
@@ -10,6 +10,7 @@ export interface Column {
   is_nullable: boolean;
   is_unique: boolean;
   is_primary_key: boolean;
+  is_generated: boolean;
   // Hard to discern in Oracle due to triggers
   has_auto_increment?: boolean;
```
Comment (aidenfoxx) [inline lib/types/column.ts:15]:
```suggestion
  has_auto_increment: boolean;
```

File: lib/types/column.ts [hunk 2]
```diff
@@ -10,6 +10,7 @@ export interface Column {
   is_nullable: boolean;
   is_unique: boolean;
   is_primary_key: boolean;
+  is_generated: boolean;
   // Hard to discern in Oracle due to triggers
```
Comment (aidenfoxx) [inline lib/types/column.ts:14]:
```suggestion
```

**Comment by Oreilles** [comment]
Welp. Don't know why, but that commit completely messed up git's diff.
The test data didn't actually change, I just added the sort function.

File: lib/utils/deep-sort.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+export default function deepSort(a: any, b: any): number | undefined {
```
Comment (aidenfoxx) [inline lib/utils/deep-sort.ts:1]: Looks like this is no longer used?-

Comment (Oreilles) [reply lib/utils/deep-sort.ts:1]: Yup, forgot to remove it!

**Comment by rijkvanzanten** [comment]
Sidenote: looks like I have to do this _every commit_ 😵

CleanShot 2021-05-17 at 12 19 21@2x

**Review by aidenfoxx — APPROVED** [review]

**Comment by rijkvanzanten** [comment]
> It's probably time for knex to use these are they're the recommanded - and SQL compliant - way. Opportunity for another PR ?

I would agree! 

@kibertoad you down for a PR that moves Oracle to use Oracle's native auto increment type instead of the trigger workaround? I realize the original code was written right around the time Oracle started implementing the auto increment type (about 7 years) ago, but seeing that's.. well.. 7 years ago now I think we can safely move over 🙂 

https://github.com/knex/knex/blob/4c21b15f1551b3f171a508937d5a3ad205865195/lib/dialects/oracle/schema/oracle-columncompiler.js#L18-L21

**Review by rijkvanzanten — APPROVED** [review]
LGTM!

**Comment by rijkvanzanten** [comment]
Ready to merge/release @Oreilles @aidenfoxx??

**Comment by Oreilles** [comment]
LGTMT 👍 It would probably be nice to add some tests on this in the close future (if not right now).

**Pull request merged**